crates/wasmtime/src/store.rs

```diff
@@ -393,6 +398,52 @@ impl Store {
             );
         }
     }
+
+    /// Returns the amount of fuel consumed by this store's execution so far.
+    ///
+    /// If fuel consumption is not enabled via [`Config::consume_fuel`] then
+    /// this function will return `None`. Also note that fuel, if enabled, must
+    /// be originally configured via [`Store::set_fuel_remaining`].
+    ///
+    /// Finally, this will only return the amount of fuel consumed since
+    /// [`Store::set_fuel_remaining`] was last called, so this does not return
+    /// the total amount of fuel consumed for the entire lifetime of this
+    /// [`Store`].
+    pub fn fuel_consumed(&self) -> Option<u64> {
+        if !self.engine().config().tunables.consume_fuel {
+            return None;
+        }
+        let consumed = unsafe { *self.inner.interrupts.fuel_consumed.get() };
+        Some(u64::try_from(self.inner.fuel_adj.get() + consumed).unwrap())
+    }
+
+    /// Updates the amount of fuel remaining in this [`Store`] available for
+    /// wasm to consume while executing.
+    ///
+    /// For this method to work fuel consumption must be enabled via
+    /// [`Config::consume_fuel`]. By default a [`Store`] starts with 0 fuel for
+    /// wasm to execute with (meaning it will immediately trap). This function
+    /// must be called for the store to have some fuel to allow WebAssembly to
+    /// execute.
+    ///
+    /// Note that at this time when fuel is entirely consumed it will cause
+    /// wasm to trap. More usages of fuel are planned for the future.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the store's [`Config`] did not have fuel
+    /// consumption enabled.
+    ///
+    /// This funtion will also panic if `remaining` is larger than
+    /// `i64::max_value()`.
+    pub fn set_fuel_remaining(&self, remaining: u64) {
+        assert!(self.engine().config().tunables.consume_fuel);
+        let remaining = i64::try_from(remaining).unwrap();
+        self.inner.fuel_adj.set(remaining);
```

Comment:
I was a tiny bit worried that a long-lived store might overflow the `i64` counter but that may not be too realistic. Do you think that'd be rare enough that we should just switch this to an add instead of a set?

Reply:
IMHO it's nicer to not have the "fuel-used value that we return is only since the last set" property -- it has the potential to become a subtle stats bug later.

Doing some quick math, a 2^63 max count, at 1B Wasm ops per second, gives us 2^33 or 8B seconds of runtime before overflow, which is ~250 years. Sometime before the year 2270 we can come back and upgrade to an `i128` :-)

Reply:
Heh good point!